server/svix-server/src/queue/redis.rs

```diff
@@ -114,39 +118,75 @@ async fn background_task(
 
     let mut pool = pool.get().await?;
 
-    // First look for delayed keys whose time is up and add them to the main qunue
-    let timestamp = Utc::now().timestamp();
-    let keys: Vec<String> = pool
-        .zrangebyscore_limit(&delayed_queue_name, 0isize, timestamp, 0isize, batch_size)
-        .await?;
+    // There is a lock on the delayed queue processing to avoid race conditions. So first try to
+    // acquire the lock setting it to a random value should it not already exist. The lock expires
+    // after five seconds in case a worker crashes while holding the lock.
 
-    if !keys.is_empty() {
-        // FIXME: needs to be a transaction
-        let keys: Vec<(String, String)> = pool
-            .zpopmin(&delayed_queue_name, keys.len() as isize)
-            .await
-            .unwrap();
-        let tasks: Vec<&str> = keys
-            .iter()
-            // All information is stored in the key in which the ID and JSON formated task
-            // are separated by a `|`. So, take the key, then take the part after the `|`
-            .map(|x| &x.0)
-            .map(|x| x.split('|').nth(1).expect("Improper key format"))
-            .collect();
-
-        // Then for each task, XADD them to ghe MAIN queue
-        let mut pipe = redis::pipe();
-        for task in tasks {
-            let _ = pipe.xadd(
-                &main_queue_name,
-                GENERATE_STREAM_ID,
-                &[(QUEUE_KV_KEY, task)],
-            );
+    // Generate a 16 character alphanumeric string
+
+    // The `thread_rng` must be contained in a smaller scope as it can not be carried across await
+    // points
+    let rand_value = {
+        let mut thread_rng = rand::thread_rng();
+        (0..16)
+            .map(|_| thread_rng.sample(Alphanumeric) as char)
+            .collect::<String>()
+    };
+
+    // SET NX PX
+    let mut cmd = redis::cmd("SET");
+    cmd.arg(DELAYED_LOCK)
+        .arg(&rand_value)
+        .arg("NX")
+        .arg("PX")
+        .arg(5000);
+    let _: () = pool.query_async(cmd).await?;
+
+    // Then check that it was set, ensuing nothing else had the lock first
+    let lock_check: String = pool.get(DELAYED_LOCK).await?;
```

Comment:
This is not how it's done, the previous command will let you know if it was set or not.